fix: ECE add-to-cart validation on product page

frosso opened the pull request:
Helps with https://github.com/Automattic/woocommerce-payments/issues/10595

#### Changes proposed in this Pull Request

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

I have been investigating server-side Store API errors triggered by the ECE, and noticed a pattern.
There are a variety of plugins that can affect the product page.
And we allow the customer to click on the ECE button from the product page itself.

However, some of the plugins affecting the product page might validate the product before it can be added to the cart: they can do that by adding additional custom fields, modifying the variation selectors, etc.

When that happens and we attempt to add the product to the cart, the Store API will fail with a `400` error - and _sometimes_ an error message (this depends on the plugin doing the validation, really).

With this PR, I'm attempting to detect when this error happens.
Now, if the customer clicks on the ECE dialog _and_ there is a server-side validation error, we're dismissing the dialog and showing an error message.
![2025-03-20 11 01 26](https://github.com/user-attachments/assets/e46326e4-21f0-44bf-b1fc-90081a7e253e)


Before, we were still showing the dialog and allowing the customer to continue interacting with the ECE dialog - without letting them know that in reality they couldn't check out because there was no product in the cart.
Before, the ECE was showing an "Invalid shipping address" message, which could have misled the customer to think that all they needed to do was to choose a different shipping address.

![2025-03-20 10 59 21](https://github.com/user-attachments/assets/dc20f372-2e31-4386-902b-3ad4148586eb)


#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

- Ensure you have the GooglePay/ApplePay buttons enabled on product pages
- Ensure you have a product with the `woo-beanie` SKU
- Add the following snippet to your store to simulate the server-side validation of a plugin:
```php
add_filter( 'woocommerce_add_to_cart_validation', function ( $valid, $product_id, $quantity ) {
	if ( wc_get_product( $product_id )->get_sku() === 'woo-beanie' ) {
		if ($quantity === 2) return true;

		return false;
	}

	if ( $quantity === 3 ) {
		return false;
	}

	return $valid;
}, 10, 3 );
```

1️⃣ 
- Visit the `woo-beanie` product page
- Click on the GooglePay/ApplePay button
- The dialog should be dismissed and an error message should appear
![2025-03-20 11 01 26](https://github.com/user-attachments/assets/05378fc7-04ba-44a3-b6c1-04766fe21d1f)

2️⃣ 
- Visit another product
- Change the quantity selector to `3`
- The GooglePay/ApplePay button should disappear
- Changing the quantity selector to another quantity should make the button re-appear, and allowing you to purchase the product
![2025-03-20 10 58 43](https://github.com/user-attachments/assets/801e4d54-58f5-4f18-a4d6-36fda31d92f5)

3️⃣ 
- Visit another product
- Click on the GooglePay/ApplePay button
- The dialog should be there, and you should be allowed to checkout the product

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

frosso committed:
fix: ECE add-to-cart valiation on product page

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `10599` or branch name `fix/ece-add-to-cart-validation-on-click` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=fix/ece-add-to-cart-validation-on-click) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 9328656373f7b6004d0a898cbdaa0dc6a2763a3a
- Build time: 2025-03-28 20:48:02 UTC

_Note: the build is updated when a new commit is pushed to this PR._

frosso committed:
better scoping

frosso commented on client/tokenized-express-checkout/__tests__/tokenized-express-checkout--product-page.test.js:
```diff
@@ -148,7 +164,7 @@ describe( 'Tokenized Express Checkout Element - Product page logic', () => {
 		} );
 
 		// triggering the `ready` event on the ECE button, to test its callback.
-		stripeElementMock[ `__triggerEvent-ready` ]( {
```
I slightly changed the implementation I made on [this PR](https://github.com/Automattic/woocommerce-payments/pull/10592) I just merged to (hopefully) make retrieving the registered event a little clearer, without the template function name 🤷

frosso commented on client/tokenized-express-checkout/__tests__/tokenized-express-checkout--product-page.test.js:
```diff
@@ -44,22 +56,22 @@ describe( 'Tokenized Express Checkout Element - Product page logic', () => {
 			displayItems: [
 				{
-					label: 'Beanie',
+					label: 'Beanie server-side',
```
I changed the naming of these labels to (hopefully) make it clearer to understand when a product is provided from the server-side hydration, instead of being provided from the Store API 🤷

frosso committed:
empty: triggering build

brettshumaker commented:
@frosso I'm getting the described behavior of `The dialog should be dismissed and an error message should appear` only in Google Chrome. In other chromium based browsers like Brave, the Google Pay dialog is not dismissed but just sits there and spins. In Safari I get the same behavior when clicking the GPay button - dialog stays but just has a spinner. When I click the Apple Pay button in Safari, the dialog stays and I get the "update shipping method" error.

<img width="600" alt="Screenshot 2025-03-20 at 11 02 15 AM" src="https://github.com/user-attachments/assets/0116eb31-37cd-456f-b339-892c05e8a901" />

Everything else seems to be working as described, the ECE buttons go away when I update the quantity to 3 (as defined in my code snippet), and I'm able to checkout normally otherwise.

frosso commented:
Thank you @brettshumaker - I knew I forgot to test something 🤦 

I reached out to Stripe here to discuss how it can be dismissed more reliably: p1742402301419669-slack-C9976E5MJ

frosso committed:
Merge branch 'develop' into fix/ece-add-to-cart-validation-on-click

frosso committed:
Merge branch 'develop' into fix/ece-add-to-cart-validation-on-click

frosso committed:
hide error message on add-to-cart error, prevent submission

frosso committed:
Merge branch 'develop' into fix/ece-add-to-cart-validation-on-click

frosso commented:
It looks like we're kinda blocked by Stripe on this front.
I added a spike to continue exploring possible solutions: https://github.com/Automattic/woocommerce-payments/issues/10647

The problem where the message is displayed on the back of the payment sheet but the payment sheet isn't dismissed persists.
However, I implemented some minor changes in 24e1cf075f666fa6477d30e1320952b059335208 so that:
- We don't show the "Invalid shipping address" message to the customers, which might mislead them to try adding additional addresses to their account
- The `confirm` handler will check if there has been an issue adding the item to the cart. If that's the case, we just dismiss the payment sheet

What do you think @brettshumaker ? Should we get this in with these workarounds in the meantime, and handle the spike as a separate thing?

brettshumaker committed:
Merge branch 'develop' into fix/ece-add-to-cart-validation-on-click

brettshumaker reviewed (approved):
Ok, I like showing the more generic message in the payment sheet. I still see Google Chrome dismissing the sheet as expected, Brave still just sits there and spins (fine), and Safari _**eventually**_ closes the sheet after about 30 seconds (without taking payment) if you try to complete the payment.

I think that these workarounds would be good to get in so we can somewhat mitigate confusion for customers. 👍
